stratum/lib/security/certificate.cc

```diff
@@ -9,132 +9,262 @@
 #include "openssl/evp.h"
 #include "openssl/pem.h"
 #include "openssl/rsa.h"
+#include "openssl/x509.h"
+#include "openssl/x509v3.h"
 #include "stratum/lib/macros.h"
 
-namespace stratum {
+/* TODO(bocon): Copied from a newer version of x509.h
+                Remove when present in the version we use.
+ */
+#ifndef X509_VERSION_3
+#define X509_VERSION_3 2
+#else
+#error "X509_VERSION_3 defined. Remove redefinition block."
+#endif
 
-using BIGNUM_ptr = std::unique_ptr<BIGNUM, decltype(&::BN_free)>;
-using BIO_ptr = std::unique_ptr<BIO, decltype(&::BIO_free)>;
-using EVP_PKEY_ptr = std::unique_ptr<EVP_PKEY, decltype(&::EVP_PKEY_free)>;
-using X509_ptr = std::unique_ptr<X509, decltype(&::X509_free)>;
+namespace stratum {
 
 // Helper functions around OpenSSL.
 namespace {
-util::StatusOr<std::string> GetRSAPrivateKeyAsString(EVP_PKEY* pkey) {
-  // Returns a reference to the underlying key; no need to free.
-  RSA* rsa = EVP_PKEY_get0_RSA(pkey);
-  CHECK_RETURN_IF_FALSE(rsa) << "Key is not an RSA key.";
 
-  BIO_ptr bio(BIO_new(BIO_s_mem()), BIO_free);
-  CHECK_RETURN_IF_FALSE(bio.get()) << "Failed to allocate string buffer.";
-  CHECK_RETURN_IF_FALSE(PEM_write_bio_RSAPrivateKey(
-      bio.get(), rsa, nullptr, nullptr, 0, nullptr, nullptr))
-      << "Failed to write private key to buffer.";
+inline bssl::UniquePtr<BIO> NewBio() {
```

Comment:
FYI: inline is only needed for functions definitions in headers so that the ODR is not violated. For helpers in anon namespaces, or in implementation files in general, it's not necessary.

https://en.cppreference.com/w/cpp/language/definition

Reply:
Good to know. Seems like these days `inline` is more about preventing ODR violations than performance, and the compiler will just inline whatever it can where it makes sense.